Integration debugging

dobosevych opened the pull request:
*Note*: Before submitting this pull request, please review our [contributing
guidelines](https://docs.celeryproject.org/en/master/contributing.html).

## Description

<!-- Please describe your pull request.

NOTE: All patches should be made against master, not a maintenance branch like
3.1, 2.5, etc.  That is unless the bug is already fixed in master, but not in
that version series.

If it fixes a bug or resolves a feature request,
be sure to link to that issue via (Fixes #4412) for example.
-->

dobosevych committed:
Integration debugging

dobosevych committed:
Integration debugging

dobosevych committed:
Integration debugging

dobosevych committed:
Commented tasks that aren't working

dobosevych committed:
Fixed test_inspect.py

dobosevych committed:
Fixed serialization test_canvas.py

dobosevych commented on t/integration/test_canvas.py:
```diff
@@ -969,7 +969,7 @@ def test_callback_called_by_group(self, manager, subtests):
             raise pytest.skip("Requires redis result backend.")
         redis_connection = get_redis_connection()
 
-        callback_msg = str(uuid.uuid4()).encode()
```
@thedrow this needs to be changed for the message.body serialization

thedrow replied:
How exactly and why?

dobosevych replied:
Signature args in callback_msg should be serializable, so
```
callback_msg = str(uuid.uuid4()).encode()
redis_key = str(uuid.uuid4())
callback = redis_echo.si(callback_msg, redis_key=redis_key)
```
,so should be
```
callback_msg = str(uuid.uuid4())
redis_key = str(uuid.uuid4())
callback = redis_echo.si(callback_msg, redis_key=redis_key)

```
as they will be passed inside as_task_v2 that would return message that would be passed to a send_task_message, then to publish (kombu) and _prepare (kombu) in messaging that would cll a serialization (kombu).
```
from kombu.serialization import dumps
s = ((42,), {}, {'callbacks': [{'task': 't.integration.tasks.redis_echo', 'args': (b'7a108737-0b0a-4acc-a32f-f8a5e6726775',), 'kwargs': {'redis_key': 'ef3bc79f-a04c-4de1-88c0-f13d9d999ca8'}, 'options': {}, 'subtask_type': None, 'immutable': True}], 'errbacks': None, 'chain': None, 'chord': None})
print(dumps(s))
```
to reproduce serialization issue

dobosevych commented on celery/worker/request.py:
```diff
@@ -154,7 +154,7 @@ def __init__(self, message, on_ack=noop,
             'routing_key': delivery_info.get('routing_key'),
             'priority': properties.get('priority'),
-            'redelivered': delivery_info.get('redelivered'),
+            'redelivered': delivery_info.get('redelivered') is True,
```
@thedrow to be sure that redelivered is always boolean

thedrow replied:
Why not simply do:
```suggestion
            'redelivered': delivery_info.get('redelivered', False),
```

If it's missing, it'll be False, otherwise, it should always be a boolean anyway.

thedrow commented on tox.ini:
```diff
@@ -42,7 +42,7 @@ deps=
 commands =
     unit: pytest --maxfail=10 -v --cov=celery --cov-report=xml --cov-report term {posargs}
-    integration: pytest -xsv t/integration {posargs}
+    integration: pytest -xsv t/integration/{posargs}
```
As we previously discussed, this is a bad idea since we'd like to pass additional positional arguments to pytest in CI which is more important than running only part of the integration suite using positional arguments. You may pass `-k <test name>` to run a specific test or file. Please revert this.

thedrow reviewed (changes requested):
Right now I have only a few minor notes. Please address them soon.

dobosevych committed:
Request fixes

dobosevych committed:
Setup full pipeline

dobosevych committed:
Setup full pipeline

dobosevych committed:
Setup full pipeline

dobosevych committed:
Setup python-package.yml

dobosevych committed:
Setup python-package.yml

thedrow commented on .github/workflows/python-package.yml:
```diff
@@ -62,13 +62,57 @@ jobs:
+      strategy:
+          fail-fast: false
+          matrix:
+              python-version: ['3.7', '3.8', '3.9']
```
We support 3.7 - 3.10. I assume you just wanted a shorter build, right?

dobosevych replied:
Added to be sure that everything is working. I also added if statement to be sure that Integration is run only, if Unit tests were successful

dobosevych committed:
Added 3.10 to integration tests

dobosevych commented on t/integration/test_tasks.py:
```diff
@@ -27,7 +27,7 @@ class test_class_based_tasks:
                                       celery_session_worker):
         task = ClassBasedAutoRetryTask()
-        celery_session_app.tasks.register(task)
+        celery_session_app.register_task(task)
```
@thedrow Integration tests were not starting for the reason of celery_class_tasks. This fixture is used inside celery.contrib.pytest that is using it in celery_session_worker. The loop for class_task in celery_class_tasks was assigning task-based signature by celery_session_app.tasks.register that is outdated interface. Here is discussion that I found about that https://stackoverflow.com/questions/41788017/register-celery-class-based-task
	
Related issues: https://github.com/celery/celery/issues/5992

dobosevych committed:
test_task.py fixed

dobosevych commented on t/integration/tasks.py:
```diff
@@ -206,7 +206,7 @@ def retry_once(self, *args, expires=60.0, max_retries=1, countdown=0.1):
                      max_retries=max_retries)
 
 
-@shared_task(bind=True, expires=60.0, max_retries=1)
```
I don't see expires in documentation of shared_task
Here are some additional confirmation of that
https://stackoverflow.com/questions/32217358/is-there-way-to-specify-expiry-when-creating-a-celery-task
expepires is possible in apply_async

dobosevych commented on t/integration/test_tasks.py:
```diff
@@ -252,13 +252,13 @@ def test_task_accepted(self, manager, sleep=1):
-    @pytest.mark.xfail(reason="Task is timeout on retry")
-    def test_task_retried_priority(self):
+    @flaky
+    def test_task_retried_priority(self, manager):
```
I suppose manager need to be added to have same task for all the task

dobosevych committed:
test_generator fixed

dobosevych commented on t/integration/test_canvas.py:
```diff
@@ -1561,12 +1561,12 @@ def test_chord_on_error(self, manager):
+            for i in range(5):
                 sleep(1)
-                if i == 2:
+                if i == 4:
```
https://github.com/celery/celery/issues/3021 has a very specific implementation. The first three elements are fetched before the execution of fourth. Fifth and fifth are yield together.
The first two fetches is happening in `_get_app` in lines
```
if tasks:
    app = tasks[0]._app
```
The third is happening in `header.tasks` in tasks, and then in the main loop inside _apply_tasks two tasks are fetched together 
```
_apply_tasks
```

dobosevych replied:
fixed

damjankuznar committed:
Ensure expiration is of float type when migrating tasks (#7385)

* Ensure expiration is of float type when migrating tasks

* Remove whitespace

* Add expiration as keyword argument to publish

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/asottile/pyupgrade: v2.31.1 → v2.32.0](https://github.com/asottile/pyupgrade/compare/v2.31.1...v2.32.0)
- [github.com/pre-commit/pre-commit-hooks: v4.1.0 → v4.2.0](https://github.com/pre-commit/pre-commit-hooks/compare/v4.1.0...v4.2.0)

dobosevych committed:
Added parametrization to test_generation

Kludex committed:
Annotate `celery/fixups` (#7399)

marksweb committed:
Fix: code block formatting for task deprecation

pyup-bot committed:
Update sphinx-click from 4.0.0 to 4.0.1

tisdall committed:
celeryproject.org links in github templates (#7442)

* update URL to contributing guidelines

* fix URL to contributing guidelines

dobosevych committed:
fixed test_generator

dobosevych committed:
Reverted encoding in test_canvas.py

simon-mazenoux committed:
fix userguide deamonizing by changing the systemd --version by systemctl --version

pyup-bot committed:
Update sphinx-click from 4.0.1 to 4.0.2

Commit:
Fix typo.

dobosevych committed:
Rollback codecov

Commit:
Retries now respect additional options.

Previously, expires and other options were not merged with
the current task's options. This commit fixes the issue.

DamianZaremba committed:
load_extension_class_names - correct module_name (#7406)

95015a changed over to using importlib rather than pkg_resources,
unfortunately the object is not exactly the same.

Attempting to start up a celery instance with `django-celery-results`
installed results in an exception during `load_extension_class_names`;

```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/celery/worker/worker.py", line 203, in start
    self.blueprint.start(self)
  File "/usr/lib/python3.10/site-packages/celery/bootsteps.py", line 112, in start
    self.on_start()
  File "/usr/lib/python3.10/site-packages/celery/apps/worker.py", line 136, in on_start
    self.emit_banner()
  File "/usr/lib/python3.10/site-packages/celery/apps/worker.py", line 170, in emit_banner
    ' \n', self.startup_info(artlines=not use_image))),
  File "/usr/lib/python3.10/site-packages/celery/apps/worker.py", line 232, in startup_info
    results=self.app.backend.as_uri(),
  File "/usr/lib/python3.10/site-packages/celery/app/base.py", line 1252, in backend
    self._local.backend = new_backend = self._get_backend()
  File "/usr/lib/python3.10/site-packages/celery/app/base.py", line 955, in _get_backend
    backend, url = backends.by_url(
  File "/usr/lib/python3.10/site-packages/celery/app/backends.py", line 69, in by_url
    return by_name(backend, loader), url
  File "/usr/lib/python3.10/site-packages/celery/app/backends.py", line 47, in by_name
    aliases.update(load_extension_class_names(extension_namespace))
  File "/usr/lib/python3.10/site-packages/celery/utils/imports.py", line 146, in load_extension_class_names
    yield ep.name, ':'.join([ep.module_name, ep.attrs[0]])
AttributeError: 'EntryPoint' object has no attribute 'module_name'
```

Move over to using the direct value should resolve this issue;

```
>>> from pkg_resources import iter_entry_points
>>> list(iter_entry_points('celery.result_backends'))[0].__dict__
{'name': 'django-cache', 'module_name': 'django_celery_results.backends', 'attrs': ('CacheBackend',), 'extras': (), 'dist': django-celery-results 2.3.0 (/usr/lib/python3.10/site-packages)}
```
vs
```
>>> from importlib.metadata import entry_points
>>> entry_points().get('celery.result_backends')[0]
EntryPoint(name='django-cache', value='django_celery_results.backends:CacheBackend', group='celery.result_backends')
```
